velox/functions/prestosql/DecimalArithmetic.cpp

```diff
@@ -0,0 +1,247 @@
+      auto flatB = args[1]->asUnchecked<FlatVector<B>>();
+      auto rawB = flatB->mutableRawValues();
+
+      R* rawResults = nullptr;
+      if (!(*result)) {
+        if (resultType->kind() == flatA->typeKind() &&
+            BaseVector::isReusableFlatVector(args[0])) {
+          rawResults = (R*)rawA;
+          *result = std::move(args[0]);
+        } else if (
+            resultType->kind() == flatB->typeKind() &&
+            BaseVector::isReusableFlatVector(args[1])) {
+          rawResults = (R*)rawB;
+          *result = std::move(args[1]);
+        }
+      }
+      if (!rawResults) {
+        rawResults = prepareResults(rows, resultType, context, result);
+      }
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], rawA[row], rawB[row], aRescale, bRescale);
+      });
+    } else {
+      exec::DecodedArgs decodedArgs(rows, args, context);
+      auto a = decodedArgs.at(0);
+      auto b = decodedArgs.at(1);
+      BaseVector::ensureWritable(rows, resultType, context->pool(), result);
+      auto rawResults =
+          (*result)->asUnchecked<FlatVector<R>>()->mutableRawValues();
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row],
+            a->valueAt<A>(row),
+            b->valueAt<B>(row),
+            aRescale,
+            bRescale);
+      });
+    }
+  }
+
+ private:
+  R* prepareResults(
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const {
+    BaseVector::ensureWritable(rows, resultType, context->pool(), result);
+    (*result)->clearNulls(rows);
+    return (*result)->asUnchecked<FlatVector<R>>()->mutableRawValues();
+  }
+
+  R* getRawResults(
+      const SelectivityVector& rows,
+      VectorPtr& flat,
+      const TypePtr& resultType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const {
+    // Check if input can be reused for results.
+    R* rawResults;
+    if (!(*result) && flat->typeKind() == resultType->kind() &&
+        BaseVector::isReusableFlatVector(flat)) {
+      auto flatValues = flat->asUnchecked<FlatVector<R>>();
+      rawResults = flatValues->mutableRawValues();
+      *result = std::move(flat);
+    } else {
+      rawResults = prepareResults(rows, resultType, context, result);
+    }
+    return rawResults;
+  }
+};
+
+class Addition {
+ public:
+  template <typename R, typename A, typename B>
+  inline static void apply(
+      R& r,
+      const A& a,
+      const B& b,
+      const uint8_t aRescale,
+      uint8_t bRescale) {
+    r.setUnscaledValue(
+        a.unscaledValue() * POWERS_OF_TEN[aRescale] +
+        b.unscaledValue() * POWERS_OF_TEN[bRescale]);
+  }
+
+  inline static uint8_t computeRescaleFactor(
+      const uint8_t fromScale,
+      const uint8_t toScale,
+      const uint8_t rScale = 0) {
```

Comment:
rScale is not used; remove

Reply:
This would be needed for division to compute rescale factors. Dividend is rescale factor is computed as
`(rs - xs + ys)`.